cpp/src/arrow/compute/exec/hash_join_node.cc

```diff
@@ -274,17 +303,83 @@ std::shared_ptr<Schema> HashJoinSchema::MakeOutputSchema(
   return std::make_shared<Schema>(std::move(fields));
 }
 
+Result<Expression> HashJoinSchema::BindFilter(Expression filter,
+                                              const Schema& left_schema,
+                                              const Schema& right_schema) {
+  if (filter.IsBound()) {
+    return std::move(filter);
+  }
```

Comment:
Maybe `if (filter.IsBound() || filter == literal(true))` and then you could get rid of the next if.